sympy/algebras/quaternion.py

```diff
@@ -321,6 +321,13 @@ def pow(self, p):
         if p == -1:
             return q.inverse()
         res = 1
+
+        if p < 0:
+            q, p = q.inverse(), -p
+
+        if (type(p) != int):
```

Comment:
Use `isinstance` to check for its type. Also, check for other integer types, like SymPy integer, `Integer`. Consider using `SYMPY_INTS` from `sympy.core.compatibility`.

Reply:
Thanks. It has been implemented. Please review.